# NovaGen Research Labs

* We develop a predictive model that classifies individuals as “healthy” or “unhealthy” based on the available health data.

# Libraries

In [23]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier,VotingClassifier
from sklearn.metrics import recall_score,classification_report,accuracy_score

# Load Dataset 

In [9]:
Data = pd.read_csv("novagen_dataset.csv")

In [12]:
Data.head()

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,...,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies,Diet_Type__Vegan,Diet_Type__Vegetarian,Blood_Group_AB,Blood_Group_B,Blood_Group_O
0,2.0,26.0,111.0,198.0,99.0,72.0,4.0,1.0,5.0,5.0,...,1,2,1,0,1,False,True,True,False,False
1,8.0,24.0,121.0,199.0,103.0,75.0,2.0,1.0,2.0,9.0,...,1,2,1,2,2,False,False,True,False,False
2,81.0,27.0,147.0,203.0,100.0,74.0,10.0,-0.0,5.0,1.0,...,2,0,0,1,0,True,False,False,False,False
3,25.0,21.0,150.0,199.0,102.0,70.0,7.0,3.0,3.0,3.0,...,1,2,1,2,0,True,False,False,True,False
4,24.0,26.0,146.0,202.0,99.0,76.0,10.0,2.0,5.0,1.0,...,2,0,2,0,2,False,True,False,True,False


In [13]:
Data.isnull().sum()

Age                      0
BMI                      0
Blood_Pressure           0
Cholesterol              0
Glucose_Level            0
Heart_Rate               0
Sleep_Hours              0
Exercise_Hours           0
Water_Intake             0
Stress_Level             0
Target                   0
Smoking                  0
Alcohol                  0
Diet                     0
MentalHealth             0
PhysicalActivity         0
MedicalHistory           0
Allergies                0
Diet_Type__Vegan         0
Diet_Type__Vegetarian    0
Blood_Group_AB           0
Blood_Group_B            0
Blood_Group_O            0
dtype: int64

* No Missing Value

In [16]:
# Data Spliting 
X = Data.drop("Target",axis=1)
y = Data["Target"]

# Train Test Split 

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2,random_state=42, stratify=y
)

* Base Model Complete

# Scaler

In [21]:
Scaler = StandardScaler()

X_scaled_train = Scaler.fit_transform(X_train)
X_scaled_test = Scaler.transform(X_test)

# Model - 1

In [28]:
# LogisticRegression Model With Regularization

log_reg = LogisticRegression(
    penalty = "l2",
    solver= "liblinear",
    max_iter=1000
)

log_reg.fit(X_scaled_train,y_train)

y_pred_lr = log_reg.predict(X_scaled_test)

# Evaluation : Recall is more important than accuracy beacuse missing a high risk patient is dangrous

print("Logistic Regression Accuracy :", accuracy_score(y_test,y_pred_lr)*100, "%")
print("Logistic Regression Recall :", recall_score(y_test,y_pred_lr)*100, "%")
print("Logistic Regression Classification Report :\n", classification_report(y_test,y_pred_lr))

Logistic Regression Accuracy : 81.41361256544502 %
Logistic Regression Recall : 82.83132530120481 %
Logistic Regression Classification Report :
               precision    recall  f1-score   support

           0       0.81      0.80      0.80       914
           1       0.82      0.83      0.82       996

    accuracy                           0.81      1910
   macro avg       0.81      0.81      0.81      1910
weighted avg       0.81      0.81      0.81      1910



# Model - 2

In [29]:
# KNN

knn = KNeighborsClassifier(
    n_neighbors=5,
    metric="euclidean"
)

knn.fit(X_scaled_train,y_train)

y_pred_knn = knn.predict(X_scaled_test)

# Evaluation

print("KNN Accuracy :", accuracy_score(y_test,y_pred_knn)*100, "%")
print("KNN Recall :", recall_score(y_test,y_pred_knn)*100, "%")
print("KNN Classification Report :\n", classification_report(y_test,y_pred_knn))

KNN Accuracy : 88.32460732984293 %
KNN Recall : 88.35341365461848 %
KNN Classification Report :
               precision    recall  f1-score   support

           0       0.87      0.88      0.88       914
           1       0.89      0.88      0.89       996

    accuracy                           0.88      1910
   macro avg       0.88      0.88      0.88      1910
weighted avg       0.88      0.88      0.88      1910



# Model - 3

In [30]:
# random Forest - Ensemble Learning

rf = RandomForestClassifier(
    n_estimators=200,
    max_depth=None,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

# Evaluation

print("KNN Accuracy :", accuracy_score(y_test,y_pred_rf)*100, "%")
print("KNN Recall :", recall_score(y_test,y_pred_rf)*100, "%")
print("KNN Classification Report :\n", classification_report(y_test,y_pred_rf))

KNN Accuracy : 93.82198952879581 %
KNN Recall : 95.88353413654619 %
KNN Classification Report :
               precision    recall  f1-score   support

           0       0.95      0.92      0.93       914
           1       0.93      0.96      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.94      0.94      1910
weighted avg       0.94      0.94      0.94      1910



# Model - 4

In [31]:
# Gradient Boosting - Ensemble Learning

gb = GradientBoostingClassifier(
    n_estimators=150,
    learning_rate=0.1,
    max_depth=3,
    random_state=42
)

gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)

# Evaluation

print("KNN Accuracy :", accuracy_score(y_test,y_pred_gb)*100, "%")
print("KNN Recall :", recall_score(y_test,y_pred_gb)*100, "%")
print("KNN Classification Report :\n", classification_report(y_test,y_pred_gb))

KNN Accuracy : 93.03664921465969 %
KNN Recall : 94.97991967871486 %
KNN Classification Report :
               precision    recall  f1-score   support

           0       0.94      0.91      0.93       914
           1       0.92      0.95      0.93       996

    accuracy                           0.93      1910
   macro avg       0.93      0.93      0.93      1910
weighted avg       0.93      0.93      0.93      1910



In [33]:
# Voting Classifier - Ensemble Learning

voting_clf = VotingClassifier(
    estimators=[
        ("lr", LogisticRegression(max_iter=1000, solver="liblinear")),
        ("knn", KNeighborsClassifier(n_neighbors=5)),
        ("rf", RandomForestClassifier(n_estimators=200, random_state=42))
    ],
    voting="soft"
)

voting_clf.fit(X_scaled_train, y_train)

y_pred_vote = voting_clf.predict(X_scaled_test)

# Evaluation

print("KNN Accuracy :", accuracy_score(y_test,y_pred_vote)*100, "%")
print("KNN Recall :", recall_score(y_test,y_pred_vote)*100, "%")
print("KNN Classification Report :\n", classification_report(y_test,y_pred_vote))

KNN Accuracy : 91.57068062827224 %
KNN Recall : 92.9718875502008 %
KNN Classification Report :
               precision    recall  f1-score   support

           0       0.92      0.90      0.91       914
           1       0.91      0.93      0.92       996

    accuracy                           0.92      1910
   macro avg       0.92      0.92      0.92      1910
weighted avg       0.92      0.92      0.92      1910



### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy of 95.88%